### DISCUSSION TITLE
Don't run actions on draft pull requests

### DISCUSSION THREAD

**Discussion by carestad** [community/community#25722] [Actions]
<p>I am trying to set up my workflow so it doesn’t run when opening/synching/reopening pull requests that are marked as draft.</p>
<p>I have found that I can use</p>
<pre><code>if: github.event.pull_request.draft == false
</code></pre>
<p>on my job, but it will still run for a few seconds, and “fail”, leaving a red X on every commit on the PR until it is marked <strong>ready for reveiw</strong>.</p>
<p>I could maybe add a dummy-job that has</p>
<pre><code>if: github.event.pull_request.draft == true
</code></pre>
<p>on it, and just echo something, but it will still run for a few seconds though.</p>
<p>My test-workfow so far looks like this:</p>
<pre><code>name: Deploy test

on:
  pull_request:
    branches: [master]
    types: [synchronize, opened, reopened, ready_for_review]

jobs:
  build:
    runs-on: ubuntu-latest
    if: github.event.pull_request.draft == false
    timeout-minutes: 10

    steps:
    - uses: actions/checkout@v1
    - name: Run all the stuff to set up my site and run tests
      run: set_up_site

  draft-build:
    runs-on: ubuntu-latest
    if: github.event.pull_request.draft == true
    timeout-minutes: 1

    steps:
    - name: Run a one-line script
      run: echo Draft PR, you are good.
</code></pre>
<p>Will those seconds be added up and billed? Is there another or better way of getting around this?</p>

**Comment by wipe2238** [comment, marked as answer]
<p>jobs.&lt;id&gt;.if is <a href="https://github.community/t/status-of-workflows-with-no-running-jobs/16782">broken</a> currently, that’s why you get “fail” instead of “skip” status.</p>

**Comment by carestad** [comment]
<p>Ah, okay. So if I just leave it as-is and they fix the issue it won’t be a problem?</p>

**Comment by wipe2238** [comment]
<p>Yeah, if you don’t mind extra notifications. I’m not touching mine waiting for a fix - beauty of betas :]</p>

**Comment by dbirks** [comment]
<p>I can confirm that this works for me now:</p>
<pre><code>if: github.event.pull_request.draft == false
</code></pre>
<p>Thanks for pointing it out!</p>

**Comment by viktorianer** [comment]
<p>Because of this, I have found a better way for myself.</p>
<pre><code>on:
  pull_request:
    types: [review_requested, ready_for_review]
</code></pre>
<p>See<br>
<a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request" rel="nofollow noopener">https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request</a><br>
<a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onevent_nametypes" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onevent_nametypes</a></p>

**Comment by julioromano** [comment]
<p><a class="mention" href="/viktorianer">@viktorianer</a> </p>
<pre><code>on:
  pull_request:
    types: [review_requested, ready_for_review]
</code></pre>
<p>This isn’t working for me. Github actions are run anyways.</p>

**Reply by Notso** [reply]
This works for me if I create two separate workflows (two separate ymls) where one only has these pull request types and another has the other types (e.g., opened, synchronize, reopened)

**Comment by julioromano** [comment]
<p><a class="mention" href="/wipe2238">@wipe2238</a> </p>
<pre><code>if: github.event.pull_request.draft == false
</code></pre>
<p>Is working but it has 2 drawbacks:</p>
<p> - The check is reported as successful even though it didn’t actually run</p>
<p> - When the PR is switched from “draft” to “ready” the CI does not run again and the check is still marked as successful so  a user is able to merge the PR without having the check actually pass. </p>

**Reply by gajus** [reply]
These are pretty major drawbacks. Just encountered this myself. Has anyone got a better suggestion?

**Reply by dalyr95** [reply]
Add the ready_for_review trigger, it does nothing by itself, if you have a trigger like push your action will still run without the if. But when you add the if, your workflow will run when you push the ready for review button

**Comment by viktorianer** [comment]
<p><a class="mention" href="/julioromano">@julioromano</a><br>
Could you provide your configuration? Because, it should work. We use it in our configuration and it is working very well.</p>

**Comment by julioromano** [comment]
<p><a class="mention" href="/viktorianer">@viktorianer</a> </p>
<pre><code>name: My CI

on:
  push:
    branches: [master]
    tags:
      - '*'
  pull_request:
    types: [review_requested, ready_for_review]
jobs:
  build:
    runs-on: ubuntu-latest
    steps:
        # omitted
</code></pre>

**Comment by viktorianer** [comment]
<p><a class="mention" href="/julioromano">@julioromano</a><br>
Thank you.</p>
<p>But I do not see my suggested solution in your configuration :roll_eyes:.</p>

**Comment by julioromano** [comment]
<p>Sorry brainfart :slight_smile: I edited it.</p>
<p>I pasted wrong code while trying other solutions.</p>

**Comment by viktorianer** [comment]
<p><a class="mention" href="/julioromano">@julioromano</a> I think it is not working because you have these two lines here, which is not what I suggested. </p>
<blockquote>
<blockquote>
<p><a class="mention" href="/julioromano">@julioromano</a> wrote:</p>
<p><a class="mention" href="/viktorianer">@viktorianer</a> </p>
<p> </p>
<p>name: My CI</p>
<p>on:<br>
push:<br>
branches: [master]<br>
…</p>
<hr>
</blockquote>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<p>You should remove them. </p>
<p>It should look like this:</p>
<pre><code>name: reviewdog
on:
  pull_request:
    types: [review_requested, ready_for_review]
jobs:
  ...:
</code></pre>

**Comment by 0gust1** [comment]
<p>I tested it, still not very practical : it would run action once after undrafting or review request, but not for further pushes on the PR (e.g. when the reviewer asked changes and the developer pushes them).</p>
<p>When your actions are running tests or lint… it’s not ideal.</p>

**Comment by viktorianer** [comment]
<p>Just invite the reviewer again (small refresh button next to the name). It is not really good, but works. Unlike other things.</p>

**Comment by natario1** [comment]
<p>It seems to me that you should just use both:</p>
<pre><code class="lang-auto">on:  
  pull_request:
    types: [opened, synchronize, reopened, ready_for_review]
jobs:
  JOB:
    if: github.event.pull_request.draft == false
</code></pre>
<p>So it won’t run while draft, but it does run when undrafting and for future updates.</p>

**Comment by lydell** [comment]
<p><a class="mention" href="/natario1">@natario1</a>’s solution works very well! Thanks!</p>
<p>I have one issue though: When I press “Ready for review”, it takes a couple of seconds for the new run to trigger. During that time the check is marked as passed. It’s enough time to easily merge the pull request. This makes it possible to accidentally bypass the branch protection rule that the check must pass before merging. :scream:</p>

**Comment by chenrui333** [comment]
<aside class="quote no-group" data-username="natario1" data-post="16" data-topic="16817">
<div class="title">
<div class="quote-controls"></div>
 natario1:</div>
<blockquote>
<pre><code class="lang-auto">on:  
  pull_request:
    types: [opened, synchronize, reopened, ready_for_review]
jobs:
  JOB:
    if: github.event.pull_request.draft == false
</code></pre>
</blockquote>
</aside>
<p>This would be a simpler solution</p>
<pre><code class="lang-auto">on:  
  pull_request:

jobs:
  JOB:
    if: github.event.pull_request.draft == false
</code></pre>

**Comment by missedone** [comment]
<p><a class="mention" href="/chenrui333">@chenrui333</a> nope, your simpler version won’t work, see the official notes:</p>
<blockquote>
<p><strong>Note:</strong> By default, a workflow only runs when a <code>pull_request</code> 's activity type is <code>opened</code> , <code>synchronize</code> , or <code>reopened</code> . To trigger workflows for more activity types, use the <code>types</code> keyword.</p>
</blockquote>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request" target="_blank" rel="noopener nofollow ugc">Events that trigger workflows - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/actions/reference/events-that-trigger-workflows</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by UnHumbleBen** [comment]
<p>You can add an additional job that fails if pull request is a draft to prevent the bypass problem:</p>
<p>Example:</p>
<pre><code class="lang-auto">
  fail_if_pull_request_is_draft:
    if: github.event.pull_request.draft == true
    runs-on: ubuntu-18.04
    steps:
    - name: Fails in order to indicate that pull request needs to be marked as ready to review and unit tests workflow needs to pass.
      run: exit 1
</code></pre>

**Comment by LeoBorai** [comment]
Using the following trigger options worked for me. You have to first open PRs as Draft but whenever you turn it _Ready for review_ CI workflows will be executed.

```yml
on:
  push:
    branches:
      - main
  pull_request:
    branches: [main]
    paths:
      - "**"
      - "!/*.md"
      - "!/**.md"
    types:
      - ready_for_review
```

**Reply by bbredewold** [reply]
This only responds to the `ready_for_review` change on the PR. You have to add conditionals to the tasks you don't want to run.

**Reply by matifali** [reply]
They would not be run on subsequent pushes i.e., the `synchronize` event

**Comment by vfilter** [comment]
This took us a while to figure out, but if you want the 
```
    types: [review_requested, ready_for_review]
```

to work, make sure your workflow has the right permission to read PRs in the repo. Without that, it won't trigger.

**Reply by h-mayorquin** [reply]
Why do you also add the `review_requested`?

**Reply by GuillermoAbadLopez** [reply]
I think, its so you can manually trigger it, whenever you want, by re-requesting review

**Comment by wolfch-elsevier** [comment]
I need the ability to to run CI on a draft PR when the PR is for changes to the GHA code itself and I don't want/need review/feedback in early stages of GHA code fixes/features.  In fact, I'm running a CI build right now on a draft PR to test a GHA change.
https://github.com/elsevier-health/hgraph/actions/runs/12675721072

**Comment by eivindjahren** [comment]
As it is now, the workarounds are too ad hoc for us. It should be possible to have the github actions not start at all as long as the PR is draft and have it start as soon as you get the ready for review event. It should also start immediately if the PR is opened  not as a draft.

**Reply by iulspop** [reply]
yes!

**Reply by alperdedeoglu** [reply]
Yes!

**Reply by justusranvier** [reply]
Nearly six years after this issue was first opened the obviously correct behavior is still impossible to achieve.

**Comment by StephenHodgson** [comment]
here's how I'm currently doing it:
```yaml
on:
  push:
    branches:
      - 'main'
      - 'development'
  pull_request:
    types: [opened, reopened, synchronize, ready_for_review]
    branches:
      - '**'
jobs:
  build:
    if: github.event.pull_request.draft == false && github.base_ref != 'main'
```

**Comment by jm-alan** [comment]
The `if ...` solution still has one moderate caveat: it only prevents the _**job**_ from running, not the _**workflow**_
This means you still spin up a runner, are charged for a minimum of 1 minute compute usage, and emit an event for all environments you consume.
The lack of ability to prevent a _**workflow**_ from running on PRs in draft is honestly pretty killer for integration and e2e testing.

**Reply by tobyward** [reply]
This isn't correct, based on this: https://github.com/orgs/community/discussions/120231#discussioncomment-9173167

According to GitHub: "You will not be billed for workflow runs that were skipped."

**Reply by ImLunaHey** [reply]
but the workflow isnt skipped.. the job is.

**Reply by jm-alan** [reply]
@tobyward that's why I specifically clarified that the ***workflow*** is ***not*** skipped.
In order for the `if ...` to even run, the workflow necessarily ***must*** run, since the statement is ***part*** of the workflow, and the `if ...` is evaluated ***on a runner***.
